drt: implement access for pins above max routing layers

eder-matheus opened the pull request.

maliberty commented on src/drt/test/top_level_term.defok:
```diff
@@ -0,0 +1,57 @@
+NETS 1 ;
+    - clk ( PIN clk ) ( clkbuf_0_clk A ) ( clkbuf_2_3__f_clk A ) ( clkbuf_2_2__f_clk A ) ( clkbuf_2_1__f_clk A ) + USE SIGNAL
+      + ROUTED met3 ( 1250 155030 ) M3M4_PR
+      NEW met4 ( 1250 155030 ) M4M5_PR
+      NEW met3 ( 1440 155030 0 ) ( 6960 * )
```
Why doesn't the m3 end at the same coordinate as the up vias?

eder-matheus replied:
DRT chooses another position for the met3 via shape, while the vias are placed using the center of the pin shape:
<img width="519" alt="image" src="https://user-images.githubusercontent.com/43450810/226762401-8a416d2e-32c2-43a2-ba8c-484875d262cd.png">

maliberty commented on src/drt/src/db/obj/frBTerm.h:
```diff
@@ -116,10 +116,14 @@ class frBTerm : public frTerm
   frBlock* block_;
   std::vector<std::unique_ptr<frBPin>> pins_;  // set later
   frNet* net_;
+  bool isAboveTopLayer_;
```
```bool isAboveTopLayer_ = false;``` avoids the need to repeat it in each ctor.

maliberty commented on src/drt/src/io/io.cpp:
```diff
@@ -904,14 +905,49 @@ void io::Parser::setBTerms(odb::dbBlock* block)
+            }
+          }
+        }
+        frCoord xl = useViaBox ? viaBox.xMin() : box->xMin();
+        frCoord yl = useViaBox ? viaBox.yMin() : box->yMin();
+        frCoord xh = useViaBox ? viaBox.xMax() : box->xMax();
+        frCoord yh = useViaBox ? viaBox.yMax() : box->yMax();
```
It seems simpler to just have a Rect to fill in rather than building it at the end:

```
Rect bbox = box->getBox();
...
if (...) {
  bbox = viaBox;
}
```

maliberty commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -976,6 +953,65 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+  std::map<int, odb::dbTechVia*> default_vias;
+  for (int i = top_layer_idx; i <= bterm_bottom_layer_idx; i++) {
+    for (odb::dbTechVia* via : vias) {
+      if (via->getBottomLayer()->getRoutingLevel() == i) {
+        default_vias[i] = via;
+        break;
+      }
```
Grabbing any via on the right layer may not work well.  At a minimum you should take a default via but even then we may need to add more checks (eg https://github.com/The-OpenROAD-Project/OpenROAD/pull/3056)

eder-matheus replied:
It's not clear how to get the default vias. To get it from DRT, it is necessary to run `initDesign` first, but my new command runs before it is initialized.

We also have this process on grt, but I think this method doesn't fit here:

```
for (odb::dbTechVia* via : vias) {
    odb::dbStringProperty* prop
        = odb::dbStringProperty::find(via, "OR_DEFAULT");

    if (prop == nullptr) {
      continue;
    } else {
      debugPrint(logger_,
                 utl::GRT,
                 "l2v_pitch",
                 1,
                 "Default via: {}.",
                 via->getConstName());
      default_vias[via->getBottomLayer()->getRoutingLevel()] = via;
    }
  }
```

eder-matheus replied:
Actually, I think it can be simplified. Perhaps we can drop this new command and just run the stack vias process inside drt, without user input. It would be simpler for users and for the implementation, as it can be executed after defining the default vias.

maliberty commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -976,6 +953,65 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+  odb::Rect pin_rect;
+  for (odb::dbBPin* bpin : bterm->getBPins()) {
+    const std::vector<odb::dbAccessPoint*>& bpin_pas = bpin->getAccessPoints();
+    access_points.insert(
+        access_points.begin(), bpin_pas.begin(), bpin_pas.end());
```
Is there a reason to front insert?  It will be slower on a vector

maliberty commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -976,6 +953,65 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+    for (const auto& ap : access_points) {
+      if (ap->getLayer()->getRoutingLevel()
+          == bterm_bottom_tech_layer->getRoutingLevel()) {
+        via_position = ap->getPoint();
```
break?

maliberty commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -976,6 +953,65 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+      }
+    }
+  } else {
+    via_position = odb::Point(pin_rect.xCenter(), pin_rect.yCenter());
```
Does this code assume you did pin access before?

eder-matheus replied:
The `else` part intends to cover the case where we don't have PA information on the bterms.

maliberty commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -976,6 +953,65 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+  // insert the vias from the top routing layer to the bterm bottom layer
+  odb::dbNet* net = bterm->getNet();
+  odb::dbWire* wire = net->getWire();
+  if (wire == nullptr) {
```
If there is an existing route then this is dangerous as you are erasing it.  It seems like an error given the intended use model.

maliberty commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -976,6 +953,65 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+  for (int layer_idx = top_layer_idx; layer_idx < bterm_bottom_layer_idx;
+       layer_idx++) {
+    odb::dbTechLayer* tech_layer = tech->findRoutingLayer(layer_idx);
+    wire_encoder.newPath(tech_layer, odb::dbWireType::ROUTED);
```
Each via doesn't need to be in a separate path as they are stacked

maliberty commented on src/drt/src/TritonRoute.tcl:
```diff
@@ -456,4 +456,51 @@ proc check_drc { args } {
 }
 
+sta::define_cmd_args "process_ports_above_top_layer" {
+    [-top_routing_layer layer]
```
This is mandatory so drop the []

maliberty commented on src/drt/src/TritonRoute.tcl:
```diff
@@ -456,4 +456,51 @@ proc check_drc { args } {
+    keys { -top_routing_layer } \
+    flags{}
+
+  if { [info exists keys(-top_routing_layer)] } {
+    set top_routing_layer $keys(-top_routing_layer)
+  } else {
+    set top_routing_layer ""
+  }
+
+  if { !($top_routing_layer eq "") } {
```
Just reduce this to a check and an error if unset

maliberty commented on src/drt/src/TritonRoute.tcl:
```diff
@@ -456,4 +456,51 @@ proc check_drc { args } {
+    set top_tech_layer [$tech findLayer $top_routing_layer]
+
+    if { $top_tech_layer == "NULL" } {
+      utl::error GRT 5 "Layer $top_routing_layer not found."
```
Give it a unique id

eder-matheus commented:
@maliberty I think I'm done with the fixes in this PR, and it's ready to merge. I could simplify the code and stack the vias during DRT, before parsing the whole design. I think I've addressed all your comments as well.

maliberty commented on src/drt/src/TritonRoute.i:
```diff
@@ -238,4 +238,12 @@ void check_drc_cmd(const char* drc_file, int x1, int y1, int x2, int y2)
   router->checkDRC(drc_file, x1, y1, x2, y2);
 }
+
+void stack_vias(odb::dbBTerm* bterm,
```
Is this just for testing or is it obsolete?  I don't any tcl caller.

eder-matheus replied:
Obsolete. I just removed it.

osamahammad21 commented on src/drt/src/TritonRoute.cpp:
```diff
@@ -986,6 +964,110 @@ void TritonRoute::checkDRC(const char* filename, int x1, int y1, int x2, int y2)
+  }
+
+  // get bterm access points
+  std::vector<odb::dbAccessPoint*> access_points;
+  odb::Rect pin_rect;
+  for (odb::dbBPin* bpin : bterm->getBPins()) {
+    const std::vector<odb::dbAccessPoint*>& bpin_pas = bpin->getAccessPoints();
+    access_points.insert(access_points.end(), bpin_pas.begin(), bpin_pas.end());
+    pin_rect = bpin->getBBox();
+  }
```
I might be wrong, so could you please modify the test case to have pin access?

eder-matheus replied:
You're right; even when running pin_access before detailed_route, this function uses the pin center to place the vias. It's ok to use the center of the pin, since the use case of this new functionality is for the bumps in the upper layers, which usually have a large area. We can drop this part of the code and always use the pin center. @maliberty, do you agree with this approach?

maliberty replied:
works for me

eder-matheus commented on src/drt/src/io/io.cpp:
```diff
@@ -904,14 +906,49 @@ void io::Parser::setBTerms(odb::dbBlock* block)
+        frLayerNum finalLayerNum = layerNum;
+        odb::Rect viaBox;
+        bool useViaBox = false;
+        if (layerNum > TOP_ROUTING_LAYER) {
+          odb::dbNet* net = term->getNet();
+          odb::dbWire* wire = net->getWire();
+          if (wire != nullptr) {
+            odb::dbWirePath path;
+            odb::dbWirePathShape pshape;
+            odb::dbWirePathItr pitr;
+            for (pitr.begin(wire); pitr.getNextPath(path);) {
+              while (pitr.getNextShape(pshape)) {
+                if (pshape.shape.isVia()) {
+                  odb::dbTechVia* via = pshape.shape.getTechVia();
+                  for (const auto& vbox : via->getBoxes()) {
+                    layerNum
+                        = tech_->name2layer[vbox->getTechLayer()->getName()]
+                              ->getLayerNum();
+                    if (layerNum == TOP_ROUTING_LAYER) {
+                      viaBox = vbox->getBox();
+                      odb::dbTransform xform;
+                      odb::Point path_origin = pshape.point;
+                      xform.setOffset({path_origin.x(), path_origin.y()});
+                      xform.setOrient(odb::dbOrientType(odb::dbOrientType::R0));
+                      xform.apply(viaBox);
+                      useViaBox = true;
+                      finalLayerNum = layerNum;
+                      termIn->setIsAboveTopLayer(true);
+                      break;
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
```
Your comment made me notice that we can have a bug here. Specifically, when we have a net with multiple pins above the max routing layer, this code is not assigning the correct via box to the pins. I believe that this part **must be** dependent on the port itself to work properly. I'm updating it now.

osamahammad21 replied:
I don't think this part is handled. you are still duplicating the same pin fig for each port that intersects with the via box. you should add the pinfig of the viabox on the top routing layer only once.

eder-matheus replied:
I think now it's handled. The code only iterates over all the bpins when the layer is below the max routing layer.

eder-matheus commented:
@maliberty @osamahammad21 Is there anything else I need to change in this PR?

maliberty commented:
@osamahammad21 please give a final review and approve if satisfied.  I have no further comments

osamahammad21 reviewed (approved):
@eder-matheus Could you please check the ci failure?

eder-matheus commented:
> @eder-matheus Could you please check the ci failure?

Not sure what caused the failure, but all the tests are passing. I think the pipeline for this run was killed before it completes.

eder-matheus commented:
@maliberty I think I need your approval too because you have requested changes.
